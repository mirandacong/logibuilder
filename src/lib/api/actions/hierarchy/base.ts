import {Action as Base} from '../action'

// tslint:disable-next-line: no-empty-interface
export interface Action extends Base {}
