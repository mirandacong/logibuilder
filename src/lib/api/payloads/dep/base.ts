import {Builder} from '@logi/base/ts/common/builder'
import {Impl} from '@logi/base/ts/common/mapped_types'

import {Payload as Base, PayloadType} from '../payload'

export interface DepPayload extends Base {
    readonly uuid: string
    readonly slice?: string
}

export abstract class DepPayloadImpl implements DepPayload {
    public uuid!: string
    public slice?: string
    public payloadType!: PayloadType
}

export class DepPayloadBuilder<T extends DepPayloadImpl,
    S extends Impl<T>> extends Builder<T, S> {
    protected get daa(): readonly string[] {
        return DepPayloadBuilder.__DAA_PROPS__
    }

    public uuid(uuid: string): this {
        this.getImpl().uuid = uuid
        return this
    }

    public slice(uuid?: string): this {
        this.getImpl().slice = uuid
        return this
    }
    protected static readonly __DAA_PROPS__: readonly string[] = [
        'uuid',
        'payloadType',
    ]
}

export function isDepPayload(obj: unknown): obj is DepPayload {
    return obj instanceof DepPayloadImpl
}

export class DepRange {
    public constructor(
        public readonly start: number,
        public readonly end: number,
    ) {}
}
