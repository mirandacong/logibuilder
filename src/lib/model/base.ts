import {Builder} from '@logi/base/ts/common/builder'
import {Impl} from '@logi/base/ts/common/mapped_types'
import {FormulaManager} from '@logi/src/lib/formula'
import {Book} from '@logi/src/lib/hierarchy/core'
import {ModifierManager} from '@logi/src/lib/modifier'
import {SourceManager} from '@logi/src/lib/source'
import {TemplateSet, TemplateSetBuilder} from '@logi/src/lib/template'

export interface Model {
    readonly book: Readonly<Book>
    readonly sourceManager: SourceManager
    readonly formulaManager: FormulaManager
    readonly modifierManager: ModifierManager
    readonly stdHeaderSet: TemplateSet
}

class ModelImpl implements Impl<Model> {
    public book!: Readonly<Book>
    public sourceManager = new SourceManager([])
    public formulaManager = new FormulaManager([])
    public modifierManager = new ModifierManager([])
    public stdHeaderSet = new TemplateSetBuilder().build()
}

export class ModelBuilder extends Builder<Model, ModelImpl> {
    public constructor(obj?: Readonly<Model>) {
        const impl = new ModelImpl()
        if (obj)
            ModelBuilder.shallowCopy(impl, obj)
        super(impl)
    }

    public book(book: Readonly<Book>): this {
        this.getImpl().book = book
        return this
    }

    public sourceManager(sourceManager: SourceManager): this {
        this.getImpl().sourceManager = sourceManager
        return this
    }

    public formulaManager(formulaManager: FormulaManager): this {
        this.getImpl().formulaManager = formulaManager
        return this
    }

    public modifierManager(modifierManager: ModifierManager): this {
        this.getImpl().modifierManager = modifierManager
        return this
    }

    public stdHeaderSet(stdHeaderSet: TemplateSet): this {
        this.getImpl().stdHeaderSet = stdHeaderSet
        return this
    }

    protected get daa(): readonly string[] {
        return ModelBuilder.__DAA_PROPS__
    }
    protected static readonly __DAA_PROPS__: readonly string[] =
        ['sourceManager', 'book']
}

export function isModel(value: unknown): value is Model {
    return value instanceof ModelImpl
}

export function assertIsModel(value: unknown): asserts value is Model {
    if (!(value instanceof ModelImpl))
        throw Error('Not a Model!')
}
