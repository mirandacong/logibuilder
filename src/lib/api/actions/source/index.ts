export {
    Action as LoadExcelAction,
    ActionBuilder as LoadExcelActionBuilder,
    isAction as isLoadExcelAction,
} from './load_excel'
export {
    Action as SetSourceAction,
    ActionBuilder as SetSourceActionBuilder,
    Argument as SetSourceArgument,
    ArgumentBuilder as SetSourceArgumentBuilder,
    isAction as isSetSourceAction,
} from './set_source'
export {
    Action as ClearDataAction,
    ActionBuilder as ClearDataActionBuilder,
    isAction as isClearDataAction,
} from './clear_data'
export {
    Action as ClearPlayGroundAction,
    ActionBuilder as ClearPlayGroundActionBuilder,
    isAction as isClearPlayGroundAction,
} from './clear_playground'
export {
    Action as SetPlaygroundSourceAction,
    ActionBuilder as SetPlaygroundSourceActionBuilder,
    isAction as isSetPlaygroundSourceAction,
} from './set_playground_source'
export {
    Action as ConfirmPlaygroundAction,
    ActionBuilder as ConfirmPlaygroundActionBuilder,
    isAction as isConfirmPlaygroundAction,
} from './confirm_playground'
export {
    Action as ResetPlaygroundAction,
    ActionBuilder as ResetPlaygroundActionBuilder,
    isAction as isResetPlaygroundAction,
} from './reset_playground_source'
