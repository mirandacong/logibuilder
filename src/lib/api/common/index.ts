export {setDcfModelHeader} from './set_dcf_model_header'
export {correctParent} from './correct_parent'
export {canChangeLevel} from './can_change_level'
export {canMoveSlicesVertically} from './can_move_slices_vertically'
export {canMoveVertically} from './can_move_vertically'
export {RefInfo, RefInfoBuilder} from './ref_info'
export {getRdepsRefInfo} from './get_rdeps_ref_info'
export {ExprBearer, ExprBearerBuilder} from './expr_bearer'
export {getUint8arrayFromBase64} from './decode_base64'
