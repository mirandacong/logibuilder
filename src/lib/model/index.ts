export {Model, ModelBuilder, assertIsModel, isModel} from './base'
export {fromJson} from './from_json'
export {toJson} from './to_json'
export {toText} from './to_text'
export {
    CloneResult,
    Config,
    ConfigBuilder,
    cloneFormula,
    cloneHierarchy,
    cloneModifier,
    cloneSource,
    cloneStdHeaderSet,
    getCloneResult,
} from './clone'
