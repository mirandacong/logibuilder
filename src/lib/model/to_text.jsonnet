local defs = import '../dsl/logi_test_data/defs.jsonnet';

local row_block2 = defs.RowBlock('rowblock', {
    modifier: '+',
    subnodes: [
        defs.Row('row1', {
            expression: '',
            labels: ['asp'],
            modifier: '---',
            is_def_scalar: false,
            slice_exprs: [
                defs.Slice('proj', ''),
            ],
        }),
        defs.Row('row2', {
            sources: [1, 2, 3, 4],
            is_def_scalar: true,
            slice_exprs: [
                defs.Slice('proj', ''),
            ],
        }),
    ],
});

local col_block2 = defs.ColBlock('cb2', {
    labels: ['label'],
    subnodes: [
        defs.Col('col1', {
            sources: [1, 2, 3, 4],
            expression: 'proj',
            modifier: '-',
            slice_exprs: [
                defs.Slice('cs1', 'proj'),
            ],
        }),
        defs.Col('col2', {
            sources: [1, 2, 3, 4],
            expression: 'proj',
            modifier: '-',
            slice_exprs: [
                defs.Slice('cs1', 'proj'),
            ],
        }),
        defs.Col('col3', {
            sources: [1, 2, 3, 4],
            expression: 'proj',
            modifier: '-',
            slice_exprs: [
                defs.Slice('cs1', 'proj'),
            ],
        }),
    ],
});

local table2 = defs.Table('table2', {
    labels: defs.Label(
        defs.Tags(['tag1', 'tag2']),
        defs.Attrs({ key: 'value' }),
    ),
    subnodes: [
        defs.RowBlock('rowblock', {
            modifier: '+',
            subnodes: [row_block2],
        }),
        defs.Row('row', {
            expression: '',
            labels: ['asp'],
            modifier: '+',
            sources: [1, 2, 3, 4, 5, 6],
            is_def_scalar: false,
            slice_exprs: [
                defs.Slice('hist', ''),
            ],
        }),
        defs.ColBlock('cb1', {
            labels: ['asp'],
            subnodes: [col_block2],
        }),
        defs.Col('col', {
            sources: [1, 2, 3, 4],
            expression: 'proj',
            modifier: '-',
            slice_exprs: [
                defs.Slice('cs1', 'proj'),
            ],
        }),
    ],
});

local title = defs.Title('title', {
    labels: defs.Label(
        defs.Tags(['tag1', 'tag2']),
        defs.Attrs({ key: 'value' }),
    ),
    subnodes: [
        defs.Table('table1', {
            labels: defs.Label(
                defs.Tags(['tag1', 'tag2']),
                defs.Attrs({ key: 'value' }),
            ),
        }),
    ],
});

defs.Book('book', {
    labels: defs.Label(
        defs.Tags(['tag1', 'tag2']),
        defs.Attrs({ key: 'value' }),
    ),
    subnodes: [
        defs.Sheet('sheet', {
            labels: defs.Label(
                defs.Tags(['tag1', 'tag2']),
                defs.Attrs({ key: 'value' }),
            ),
            subnodes: [
                title,
                table2,
            ],
        }),
    ],
})
