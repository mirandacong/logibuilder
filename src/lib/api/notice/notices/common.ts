import {Builder} from '@logi/base/ts/common/builder'
import {Impl} from '@logi/base/ts/common/mapped_types'

import {Message, MessageBuilder, Type} from '../message'
import {Notice} from '../notice'

export interface CommonNotice extends Notice {
    readonly actionType: number
    readonly message: string
    readonly type: Type
    getMessage(): Message
}

class CommonNoticeImpl implements Impl<CommonNotice> {
    public actionType!: number
    public message!: string
    public type!: Type

    // tslint:disable-next-line: no-unnecessary-method-declaration
    public getMessage(): Message {
        return new MessageBuilder().main(this.message).type(this.type).build()
    }
}

export class CommonNoticeBuilder extends
    Builder<CommonNotice, CommonNoticeImpl> {
    public constructor(obj?: Readonly<CommonNotice>) {
        const impl = new CommonNoticeImpl()
        if (obj)
            CommonNoticeBuilder.shallowCopy(impl, obj)
        super(impl)
    }

    public actionType(actionType: number): this {
        this.getImpl().actionType = actionType
        return this
    }

    public message(message: string): this {
        this.getImpl().message = message
        return this
    }

    public type(type: Type): this {
        this.getImpl().type = type
        return this
    }

    protected get daa(): readonly string[] {
        return CommonNoticeBuilder.__DAA_PROPS__
    }
    protected static readonly __DAA_PROPS__: readonly string[] = [
        'actionType',
        'message',
        'type',
    ]
}

export function isCommonNotice(value: unknown): value is CommonNotice {
    return value instanceof CommonNoticeImpl
}

export function assertIsCommonNotice(
    value: unknown,
): asserts value is CommonNotice {
    if (!(value instanceof CommonNoticeImpl))
        throw Error('Not a CommonNotice!')
}
