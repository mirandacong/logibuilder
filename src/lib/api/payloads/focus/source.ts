import {Builder} from '@logi/base/ts/common/builder'
import {Impl} from '@logi/base/ts/common/mapped_types'

import {Payload as Base, PayloadType} from '../payload'

export interface FocusSourcePayload extends Base {
    readonly row: string
    readonly col: string
}

class FocusSourcePayloadImpl implements Impl<FocusSourcePayload> {
    public row!: string
    public col!: string
    public payloadType = PayloadType.FOCUS_SOURCE
}

export class FocusSourcePayloadBuilder extends
    Builder<FocusSourcePayload, FocusSourcePayloadImpl> {
    public constructor(obj?: Readonly<FocusSourcePayload>) {
        const impl = new FocusSourcePayloadImpl()
        if (obj)
            FocusSourcePayloadBuilder.shallowCopy(impl, obj)
        super(impl)
    }

    public row(row: string): this {
        this.getImpl().row = row
        return this
    }

    public col(col: string): this {
        this.getImpl().col = col
        return this
    }

    protected get daa(): readonly string[] {
        return FocusSourcePayloadBuilder.__DAA_PROPS__
    }
    protected static readonly __DAA_PROPS__: readonly string[] = [
        'row',
        'col',
    ]
}

export function isFocusSourcePayload(
    value: unknown,
): value is FocusSourcePayload {
    return value instanceof FocusSourcePayloadImpl
}

export function assertIsFocusSourcePayload(
    value: unknown,
): asserts value is FocusSourcePayload {
    if (!(value instanceof FocusSourcePayloadImpl))
        throw Error('Not a FocusSourcePayload!')
}
