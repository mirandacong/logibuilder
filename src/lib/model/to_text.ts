import {Builder} from '@logi/base/ts/common/builder'
import {Impl} from '@logi/base/ts/common/mapped_types'
import {postOrderWalk2} from '@logi/base/ts/common/walk_utils'
import {
    Column,
    isBook,
    isColumn,
    isColumnBlock,
    isRow,
    isRowBlock,
    isSheet,
    isTable,
    isTitle,
    Node,
    NodeType,
    Row,
    RowBlock,
    SliceExpr,
    Table,
} from '@logi/src/lib/hierarchy/core'
import {SourceManager} from '@logi/src/lib/source'
import {safeDump} from 'js-yaml'

import {Model} from './base'

// tslint:disable-next-line: unknown-instead-of-any
type TextObj = any
type ToTextMap = Map<Readonly<Node>, TextObj>
const INDENT = 4
const NEW_LINE = '\n'
const START = '---'

export function toText(model: Readonly<Model>): string {
    const loadMap: ToTextMap = new Map()
    const walkData = new WalkDataBuilder()
        .sourceManager(model.sourceManager)
        .map(loadMap)
        .build()
    const node = model.book
    postOrderWalk2(node, loadVisitor, getSubnodes, walkData)
    const obj = loadMap.get(node)
    return getContent(safeDump(obj, {indent: INDENT}))
}

function getContent(str: string): string {
    return [START, str].join(NEW_LINE)
}

function getSubnodes(node: Readonly<Node>): readonly Readonly<Node>[] {
    const children: Readonly<Node>[] = []
    if (isBook(node))
        children.push(...node.sheets)
    else if (isSheet(node) || isTitle(node) || isRowBlock(node)
        || isColumnBlock(node))
        children.push(...node.tree)
    else if (isTable(node))
        children.push(...node.rows, ...node.cols)
    return children
}

// tslint:disable-next-line: max-func-body-length cyclomatic-complexity
function loadVisitor(
    node: Readonly<Node>,
    walkData: WalkData,
): readonly TextObj[] {
    // tslint:disable-next-line: no-object-literal-type-assertion
    const textObj = {} as TextObj
    const map = walkData.map
    switch (node.nodetype) {
    case NodeType.BOOK: {
        const labels = getLabels(node)
        const value = labels.length === 0 ? `${node.name}` :
            `${node.name}[${labels}]`
        textObj.Book = value
        setSubnodes(node, textObj, map)
        break
    }
    case NodeType.SHEET: {
        const labels = getLabels(node)
        const value = labels.length === 0 ? `${node.name}` :
            `${node.name}[${labels}]`
        textObj.Sheet = value
        setSubnodes(node, textObj, map)
        break
    }
    case NodeType.TITLE: {
        const labels = getLabels(node)
        const value = labels.length === 0 ? `${node.name}` :
            `${node.name}[${labels}]`
        textObj.Title = value
        setSubnodes(node, textObj, map)
        break
    }
    case NodeType.TABLE: {
        const n = node as Table
        textObj.header_stub = n.headerStub
        const labels = getLabels(n)
        const value = labels.length === 0 ? `${n.name}` :
            `${n.name}[${labels}]`
        textObj.Table = value
        const cols: TextObj[] = []
        const rows: TextObj[] = []
        getSubnodes(n).forEach((sub: Readonly<Node>): void => {
            const subString = map.get(sub)
            if (subString === undefined)
                return
            if (isRowBlock(sub) || isRow(sub))
                rows.push(subString)
            if (isColumnBlock(sub) || isColumn(sub))
                cols.push(subString)
        })
        textObj.rows = rows
        textObj.cols = cols
        break
    }
    case NodeType.COLUMN_BLOCK: {
        const labels = getLabels(node)
        const value = (labels.length === 0) ? `${node.name}` :
            `${node.name}[${labels}]`
        textObj.ColumnBlock = value
        setSubnodes(node, textObj, map)
        break
    }
    case NodeType.ROW_BLOCK: {
        const n = node as RowBlock
        const labels = getLabels(n)
        const value = labels.length === 0 ? `${n.name}` :
            `${n.name}[${labels}]`
        textObj.RowBlock = value
        setSubnodes(node, textObj, map)
        break
    }
    case NodeType.ROW: {
        const n = node as Row
        const labels = getLabels(node)
        const name = labels.length === 0 ? `${n.name}` :
            `${n.name}[${labels}]`
        Reflect.set(textObj, name, n.expression)
        setSources(n, textObj, walkData.sourceManager)
        const slices = n.sliceExprs.map(sliceToText)
        textObj.slice_exprs = slices
        break
    }
    case NodeType.COLUMN: {
        const n = node as Column
        const labels = getLabels(node)
        const name = labels.length === 0 ? `${n.name}` :
            `${n.name}[${labels}]`
        Reflect.set(textObj, name, n.expression)
        const slices = n.sliceExprs.map(sliceToText)
        textObj.slice_exprs = slices
        break
    }
    default:
    }
    map.set(node, textObj)
    return []
}

function getLabels(node: Readonly<Node>): readonly string[] {
    return node.labels
}

function setSubnodes(
    node: Readonly<Node>,
    textObj: TextObj,
    map: ToTextMap,
): void {
    const subs: TextObj[] = []
    getSubnodes(node).forEach((sub: Readonly<Node>): void => {
        const subString = map.get(sub)
        if (subString !== undefined)
            subs.push(subString)
    })
    textObj.subnodes = subs
}

function setSources(
    node: Readonly<Row>,
    textObj: TextObj,
    manager: SourceManager,
): void {
    const table = node.getTable()
    if (!isTable(table))
        return
    const cols = table.getLeafCols()
    const result: (number | string)[] = []
    let nullCount = 0
    cols.forEach((col: Readonly<Column>): void => {
        const source = manager.getSource(node.uuid, col.uuid)
        if (source !== undefined && source.value !== '') {
            result.push(source.value)
            return
        }
        result.push('null')
        nullCount += 1
    })
    if (nullCount === result.length)
        return
    textObj.sources = result
}

function sliceToText(slice: SliceExpr): TextObj {
    const slices = {}
    const name = slice.name
    const expression = slice.expression
    Reflect.set(slices, name, expression)
    return slices
}

interface WalkData {
    readonly map: ToTextMap
    readonly sourceManager: SourceManager
}

class WalkDataImpl implements Impl<WalkData> {
    public map!: ToTextMap
    public sourceManager!: SourceManager
}

export class WalkDataBuilder extends Builder<WalkData, WalkDataImpl> {
    public constructor(obj?: Readonly<WalkData>) {
        const impl = new WalkDataImpl()
        if (obj)
            WalkDataBuilder.shallowCopy(impl, obj)
        super(impl)
    }

    public map(map: ToTextMap): this {
        this.getImpl().map = map
        return this
    }

    public sourceManager(sourceManager: SourceManager): this {
        this.getImpl().sourceManager = sourceManager
        return this
    }

    protected get daa(): readonly string[] {
        return WalkDataBuilder.__DAA_PROPS__
    }
}

export function isWalkData(value: unknown): value is WalkData {
    return value instanceof WalkDataImpl
}

export function assertIsWalkData(value: unknown): asserts value is WalkData {
    if (!(value instanceof WalkDataImpl))
        throw Error('Not a WalkData!')
}
