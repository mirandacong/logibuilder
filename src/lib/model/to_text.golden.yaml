---
Book: book[tag1,tag2,key:value]
subnodes:
- Sheet: sheet[tag1,tag2,key:value]
  subnodes:
  - Title: title[tag1,tag2,key:value]
    subnodes:
    - header_stub: ''
      Table: table1[tag1,tag2,key:value]
      rows: []
      cols: []
  - header_stub: ''
    Table: table2[tag1,tag2,key:value]
    rows:
    - RowBlock: +rowblock
      subnodes:
      - RowBlock: +rowblock
        subnodes:
        - '---row1[asp]': ''
          slice_exprs:
          - proj: ''
        - row2: ''
          sources:
          - 1
          - 2
          - 3
          - 4
          slice_exprs:
          - proj: ''
    - +row[asp]: ''
      sources:
      - 1
      - 2
      - 3
      - 4
      - 5
      - 6
      slice_exprs:
      - hist: ''
    cols:
    - ColumnBlock: cb1[asp]
      subnodes:
      - ColumnBlock: cb2[label]
        subnodes:
        - -col1: proj
          sources:
          - 1
          - 2
          - 3
          - 4
          slice_exprs:
          - cs1: proj
        - -col2: proj
          sources:
          - 1
          - 2
          - 3
          - 4
          slice_exprs:
          - cs1: proj
        - -col3: proj
          sources:
          - 1
          - 2
          - 3
          - 4
          slice_exprs:
          - cs1: proj
    - -col: proj
      sources:
      - 1
      - 2
      - 3
      - 4
      slice_exprs:
      - cs1: proj
